copybookreader-plugins/src/main/java/co/cask/hydrator/plugin/batch/CopyBookSource.java

```diff
@@ -0,0 +1,283 @@
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+package co.cask.hydrator.plugin.batch;
+
+import co.cask.cdap.api.annotation.Description;
+import co.cask.cdap.api.annotation.Name;
+import co.cask.cdap.api.annotation.Plugin;
+import co.cask.cdap.api.data.batch.Input;
+import co.cask.cdap.api.data.format.StructuredRecord;
+import co.cask.cdap.api.data.schema.Schema;
+import co.cask.cdap.api.dataset.lib.KeyValue;
+import co.cask.cdap.etl.api.Emitter;
+import co.cask.cdap.etl.api.PipelineConfigurer;
+import co.cask.cdap.etl.api.batch.BatchRuntimeContext;
+import co.cask.cdap.etl.api.batch.BatchSource;
+import co.cask.cdap.etl.api.batch.BatchSourceContext;
+import co.cask.hydrator.common.ReferencePluginConfig;
+import co.cask.hydrator.common.SourceInputFormatProvider;
+import co.cask.hydrator.common.batch.JobUtils;
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.base.Preconditions;
+import com.google.common.base.Strings;
+import com.google.common.collect.Lists;
+import net.sf.JRecord.Common.AbstractFieldValue;
+import org.apache.avro.reflect.Nullable;
+import org.apache.hadoop.conf.Configuration;
+import org.apache.hadoop.io.LongWritable;
+import org.apache.hadoop.mapreduce.Job;
+
+import java.io.IOException;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+
+/**
+ * Batch source to poll fixed-length flat files that can be parsed using a COBOL copybook.
+ * <p>
+ * The plugin will accept the copybook contents in a textbox and a binary data file.
+ * It produces structured records based on the schema as defined either by the copybook contents or the user.
+ * <p>
+ * For this first implementation, it will only accept binary fixed-length flat files without any nesting.
+ */
+@Plugin(type = BatchSource.PLUGIN_TYPE)
+@Name("CopybookReader")
+@Description("Batch Source to read COBOL Copybook fixed-length flat files")
+public class CopybookSource extends BatchSource<LongWritable, Map<String, AbstractFieldValue>, StructuredRecord> {
+
+  private final CopybookSourceConfig config;
+  private Schema outputSchema;
+
+  public CopybookSource(CopybookSourceConfig copybookConfig) {
+    this.config = copybookConfig;
+  }
+
+
+  @Override
+  public void configurePipeline(PipelineConfigurer pipelineConfigurer) {
+    super.configurePipeline(pipelineConfigurer);
+    config.validate();
+    pipelineConfigurer.getStageConfigurer().setOutputSchema(config.parseSchema());
+
+  }
+
+  @Override
+  public void initialize(BatchRuntimeContext context) throws Exception {
+    super.initialize(context);
+    if (config.schema != null) {
+      outputSchema = config.parseSchema();
+    }
+  }
+
+  @Override
+  public void prepareRun(BatchSourceContext context) throws IOException {
+    Job job = JobUtils.createInstance();
+    Configuration conf = job.getConfiguration();
+    conf.set(CopybookConstants.COPYBOOK_INPUTFORMAT_CBL_CONTENTS, config.copybookContents);
```

Comment:
Should move setting of configuration into the input format class. Most input formats contain public static methods where they will set the relevant conf settings. This makes it more re-usable since logic for configuring the input format is in the format itself and not in the classes that call it. For example, if another plugin needs to use the input format.